provider/provider.go

```diff
@@ -62,8 +64,17 @@ func (p configurableProvider) Configure(
 	ctx context.Context,
 	request *rpc.ConfigureRequest,
 ) (*rpc.ConfigureResponse, error) {
-	schema := internal.Schema(ctx, Version)
-	ce := deprecated.New(schema.Config)
+	r, err := p.GetSchema(ctx, &rpc.GetSchemaRequest{Version: 0})
+	if err != nil {
+		return nil, err
+	}
+	spec := schema.PackageSpec{}
+	err = json.Unmarshal([]byte(r.Schema), &spec)
+	if err != nil {
+		return nil, err
+	}
+
+	ce := deprecated.New(spec.Config)
```

Comment:
@iwahbe FWIW my E2E tests would panic with the previous logic due to missing `RunInfo`. Going through the RPC handler instead of `provider.GetSchema` directly fixed things 🤷 

This path is pretty weird/convoluted so it could be I was just doing something silly.

```
panic: interface conversion: interface {} is nil, not provider.RunInfo

goroutine 22 [running]:

github.com/pulumi/pulumi-go-provider.GetRunInfo(...)

	/Users/bryce/go/pkg/mod/github.com/pulumi/pulumi-go-provider@v0.16.1-0.20240426163030-72606b1e6fb7/provider.go:521

github.com/pulumi/pulumi-go-provider/middleware/schema.(*state).generateSchema(_, {_, _})

	/Users/bryce/go/pkg/mod/github.com/pulumi/pulumi-go-provider@v0.16.1-0.20240426163030-72606b1e6fb7/middleware/schema/schema.go:283 +0x774

github.com/pulumi/pulumi-go-provider/middleware/schema.(*state).GetSchema(0x14000798900, {0x10641b138, 0x14000f8b8f0}, {0x14000a32038?})

	/Users/bryce/go/pkg/mod/github.com/pulumi/pulumi-go-provider@v0.16.1-0.20240426163030-72606b1e6fb7/middleware/schema/schema.go:185 +0x60

github.com/pulumi/pulumi-go-provider/middleware/context.Wrap.delegateIO[...].func1({0x140002aa240?})

	/Users/bryce/go/pkg/mod/github.com/pulumi/pulumi-go-provider@v0.16.1-0.20240426163030-72606b1e6fb7/middleware/context/context.go:50 +0x50

github.com/pulumi/pulumi-go-provider/middleware/cancel.Wrap.setCancel2[...].func7({0x0?})

	/Users/bryce/go/pkg/mod/github.com/pulumi/pulumi-go-provider@v0.16.1-0.20240426163030-72606b1e6fb7/middleware/cancel/cancel.go:133 +0xa0

github.com/pulumi/pulumi-go-provider.GetSchema({_, _}, {_, _}, {_, _}, {0x14000f8b6e0, 0x14000f86880, 0x14000f8b710, 0x14000f8b740, ...})

	/Users/bryce/go/pkg/mod/github.com/pulumi/pulumi-go-provider@v0.16.1-0.20240426163030-72606b1e6fb7/provider.go:480 +0x120

github.com/pulumi/pulumi-docker-build/provider/internal.Schema({_, _}, {_, _})

	/Users/bryce/src/pulumi-docker-build/provider/internal/provider.go:132 +0x108

github.com/pulumi/pulumi-docker-build/provider.configurableProvider.Configure({{0x10643f8f0?, 0x14000c84480?}}, {0x10641b138, 0x14000f8b0b0}, 0x14000e3c140)

	/Users/bryce/src/pulumi-docker-build/provider/provider.go:65 +0x60

github.com/pulumi/pulumi/sdk/v3/proto/go._ResourceProvider_Configure_Handler.func1({0x10641b138?, 0x14000f8b0b0?}, {0x1061f4e80?, 0x14000e3c140?})

	/Users/bryce/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.113.3/proto/go/provider_grpc.pb.go:482 +0xd0

github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1({0x10641b138, 0x14000f8aba0}, {0x1061f4e80, 0x14000e3c140}, 0x14000e6ac60, 0x140007429c0)

	/Users/bryce/go/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20180507213350-8e809c8a8645/go/otgrpc/server.go:57 +0x2e4

github.com/pulumi/pulumi/sdk/v3/proto/go._ResourceProvider_Configure_Handler({0x1062ef5e0, 0x14000a1e180}, {0x10641b138, 0x14000f8aba0}, 0x14000f86380, 0x14000aa0300)

	/Users/bryce/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.113.3/proto/go/provider_grpc.pb.go:484 +0x148

google.golang.org/grpc.(*Server).processUnaryRPC(0x14000e24000, {0x10641b138, 0x14000f8ab10}, {0x106431180, 0x140001924e0}, 0x14000dab680, 0x1400085cdb0, 0x107e69be8, 0x0)

	/Users/bryce/go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1386 +0xb58

google.golang.org/grpc.(*Server).handleStream(0x14000e24000, {0x106431180, 0x140001924e0}, 0x14000dab680)

	/Users/bryce/go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1797 +0xb10

google.golang.org/grpc.(*Server).serveStreams.func2.1()

	/Users/bryce/go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1027 +0x8c

created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 85

	/Users/bryce/go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1038 +0x13c

error: error reading from server: EOF
```

Reply:
Unfortunately, I think the behavior you are seeing makes sense here. The RPC handler from `pulumi-go-provider` handles injecting `RunInfo`, which is 100% necessary for `GetSchema` to complete. When the method is directly called, it doesn't have access to RunInfo and can't complete.

This solution "works", for some definition of works. My recommendation would be to rewrite `configurableProvider` as `pulumi-go-provider` level middleware. Then it will be safe to just call `GetSchema` since `RunInfo` will have already been injected.

Reply:
Maybe something like this: https://github.com/pulumi/pulumi-docker-build/compare/blampe/upgrade-go-provider...iwahbe/upgrade-go-provider.

Reply:
Pulled that in and got tests passing again. Worth mentioning part of the reason this was so wonky was because the config encoding stuff was vendored directly from the bridge. I don't love that we have to fork it a bit here, but it's a good reminder to dust off https://github.com/pulumi/pulumi/pull/15032 so we can stop dealing with this.